### PR TITLE
llm: gptoss

### PR DESCRIPTION
**Pull request by b1tg**
gpt-oss shares a few components (yarn, sliding window, gate/up clamp) with deepseek v4, plan is to get this merged first

### CODE REVIEW

File: tinygrad/llm/model.py [hunk 1]
```diff
@@ -56,23 +68,32 @@ class TransformerConfig:
   rope_theta: float
   rope_dim: int
   v_head_dim: int
+  yarn: tuple[float, int, float, float]|None = None
   max_context: int = 0
   qk_norm: int = 0
   num_experts: int = 0
   num_experts_per_tok: int = 0
   norm_topk_prob: bool = False
+  # 1: softmax, 2: sigmoid, 3: softmax over top-k logits, 4: sqrt-softplus
```
Comment (geohot) [inline tinygrad/llm/model.py]: Don't do this, this should be an enum

File: tinygrad/llm/model.py [hunk 2]
```diff
@@ -56,23 +68,32 @@ class TransformerConfig:
   rope_theta: float
   rope_dim: int
   v_head_dim: int
+  yarn: tuple[float, int, float, float]|None = None
```
Comment (geohot) [inline tinygrad/llm/model.py]: What is this? It at least needs a comment

File: tinygrad/llm/model.py [hunk 3]
```diff
@@ -56,23 +68,32 @@ class TransformerConfig:
...
+  attn_sinks: bool = False
   ssm: SSMConfig|None = None
   shared_expert_gate: bool = True
   leading_dense_blocks: int = 0
   dense_hidden_dim: int = 0
   routed_scaling_factor: float = 1.0
   qkv_bias: bool = False
   expert_bias: bool = False
+  expert_proj_bias: bool = False
+  oai_swiglu: bool = False
```
Comment (geohot) [inline tinygrad/llm/model.py]: Is this the realy name in the literature?

Comment (b1tg) [reply tinygrad/llm/model.py]: it's a specialized variant of the SwiGLU used by gptoss/minimax-m3, not mentioned in gguf

**Comment by geohot** [comment]
Was this tested with real gpt-oss? What GGUF did you use?

Kimi says this.

 Bugs (both verified against the real GGUF metadata)

 1. Wrong MoE routing for gpt-oss. The GGUF has no gpt-oss.expert_gating_func or gpt-oss.expert_weights_norm keys, so the config defaults to ExpertGating.SOFTMAX with norm_topk_prob=False: softmax over all 32 experts, then top-4 gather without renormalization. But HF (GptOssTopKRouter) and llama.cpp both do top-k first, then softmax over the
 selected k — that's tinygrad's SOFTMAX_WEIGHT (router bias is inside the Linear, so it correctly participates). With the current default, routing weights sum to <1 and vary per token. Fix: derive SOFTMAX_WEIGHT for arch == 'gpt-oss'.

 2. Sliding window applied to all 24 layers. The GGUF has gpt-oss.attention.sliding_window=128 but no sliding_window_pattern key (llama.cpp hardcodes the alternation per-arch), so sliding_window_pattern defaults to 0 → every layer is sliding-window. HF layer_types = [sliding, full, sliding, full, …]. Note the PR's parity convention ((i+1) %
 pattern == 0 → full) is right for pattern=2 (layer 0 sliding). Fix: default the pattern to 2 for gpt-oss.

 3. gpt-4o preset uses the wrong pretokenizer regex. Verified with tiktoken's o200k: it splits case runs — "HelloWorld"→["Hello","World"], "GPTs"→["GPT","s"], "iPhone"→["i","Phone"] — while SimpleTokenizer's cl100k-style regex keeps them whole. BPE over wrong pretokens ≠ o200k tokens. gpt-oss needs the gpt-4o regex variant, not just the name
 added to the allowlist.

**Comment by b1tg** [comment]
I was testing with the [unsloth gguf](https://huggingface.co/unsloth/gpt-oss-20b-GGUF/blob/main/gpt-oss-20b-Q4_K_M.gguf). I did cli/serve tests early on, but later only using `--benchmark` to re-verify. I accidentally lost the two gptoss defaults during cleanup. I'll be more careful next time. 

Also, the pretokenizer regex might not be fully correct for other models either, I'll look into it.

**Comment by geohot** [comment]
GPT 5.6 Sol's thoughts (this is how I review all my PRs too)

 Findings

 1. [P1] Preserve GPT-OSS’s fractional YaRN boundaries —
    tinygrad/llm/model.py:29-30
    GPT-OSS specifies truncate: false; its canonical implementation uses
    fractional bounds (~8.093 and ~17.398). Flooring/ceiling them to 8/18
    changes RoPE frequencies and produces increasingly incorrect attention
    logits at longer positions. Add a YaRN truncation option and disable it
    for GPT-OSS, or retain fractional bounds.

 2. [P1] Don’t apply GPT-OSS YaRN scaling to every YaRN model —
    tinygrad/llm/model.py:27, tinygrad/llm/model.py:465-468
    Every GGUF declaring YaRN now gets 1 + 0.1*log(factor) applied to RoPE.
    DeepSeek2/R1 instead uses yarn_log_multiplier and separates rotary
    scaling from the global attention-logit scale. For DeepSeek-R1 this
    code scales only the rotary Q/K dimensions while leaving non-RoPE MLA
    dimensions unscaled, producing incorrect attention. Parse the scaling
    metadata and support MLA’s attention scale, or initially restrict this
    YaRN path to GPT-OSS.

**Comment by b1tg** [comment]
1. The rounding is intentional, this implementation follows [llama.cpp](https://github.com/ggml-org/llama.cpp/blob/a3b1effcda84caeb180427b1346d0212841418f6/ggml/src/ggml.c#L4370-L4384)
2. This is currently the first model for which we support yarn, so this implementation targets its behavior, we can add model-specific scaling when we support other models
